addon/scheduler.ts

```diff
@@ -26,19 +33,51 @@ let _whenRoutePainted: Promise<any>;
 let _whenRoutePaintedScheduleFn: Function;
 let _whenRouteIdle: Promise<any>;
 let _whenRouteIdleScheduleFn: Function;
-let _activeScheduledTasks: number = 0;
 const CAPABILITIES: Capabilities = {
   requestAnimationFrameEnabled: typeof requestAnimationFrame === 'function',
   requestIdleCallbackEnabled: typeof requestIdleCallback === 'function',
+  performanceObserverEnabled: typeof PerformanceObserver === 'function',
 };
+export const DEFAULT_MARK_NAME: string = 'routeIdle';
+const DEFAULT_OPTIONS: SchedulerOptions = {
+  markName: DEFAULT_MARK_NAME,
+  emitMarks: true,
+};
+
 let _capabilities = CAPABILITIES;
+let schedulerOptions: SchedulerOptions = DEFAULT_OPTIONS;
 
 export const USE_REQUEST_IDLE_CALLBACK: boolean = true;
 export const SIMPLE_CALLBACK = (callback: Function) => callback();
 const IS_FASTBOOT = typeof (<any>window).FastBoot !== 'undefined';
+const waiter = buildWaiter('ember-app-scheduler-waiter');
+let _emitMarks: Function = emitMarks;
 
 reset();
 
+function installObserver() {
+  let observer = new PerformanceObserver(list => {
+    let entries = list.getEntriesByName(schedulerOptions.markName);
+
+    if (entries.length > 0) {
+      _whenRouteDidChange.resolve();
+    }
+  });
+  observer.observe({ entryTypes: ['mark'] });
+}
+
+export function emitMarks() {
+  let token = waiter.beginAsync();
+  schedule('afterRender', null, () => {
+    requestAnimationFrame(() => {
+      requestAnimationFrame(() => {
+        performance.mark(schedulerOptions.markName);
+        waiter.endAsync(token);
```

Comment:
I'm not 100% sure this is the right timing for endAsync. There is likely to be a gap in time between this `performance.mark` call and the ultimate `_whenRouteDidChange.resolve()` call that we will miss...

Reply:
So you could be right, and the way to ensure this is correct is to hoist this to module scope and track the async there. We can then `endAsync` once the resolve is completed, as we do below. Thoughts?

Reply:
This is slightly more tricky than it seems. The more appropriate spot to call `endAsync` would be in the `PerformanceObserver`, just before it calls resolve on the deferred promise. The tricky piece is that we're not guaranteed, in tests, to can _this_ `emitMark` function, as we override it for testing purposes (if we don't, our tests that validate whether `requestAnimationFrame` is invoked correctly will over-fire their assertions). 

I can of course make this work, but it adds to the complexity.